### PR TITLE
cloud/ovirt: set storage domain when creating vm

### PR DESCRIPTION
**Pull request by tareqalayan**
##### ISSUE TYPE

 - Feature Pull Request
 - New Module Pull Request
 - Bugfix Pull Request
 - Docs Pull Request

##### ANSIBLE VERSION

```

```

##### SUMMARY

```

```

### CODE REVIEW

**Comment by tareqalayan** [comment]
@machacekondra please review thanks

File: lib/ansible/modules/cloud/ovirt/ovirt_vms.py [hunk 1]
```diff
@@ -444,8 +447,36 @@ def __get_template_with_version(self):
 
         return template
 
+    def __get_storage_domain(self):
```
Comment (machacekondra) [inline lib/ansible/modules/cloud/ovirt/ovirt_vms.py]: You may use `search_by_name()' method:

```
search_by_name(
    self._connection.system_service().storage_domains_service(),
    self._module.params['storage_domain']
)
```

https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/ovirt.py#L221

File: lib/ansible/modules/cloud/ovirt/ovirt_vms.py [hunk 2]
```diff
@@ -444,8 +447,36 @@ def __get_template_with_version(self):
...
+
+        return storage_domain
+
+    def __get_disk_attachments(self, template):
+        template_service = self._connection.system_service().templates_service().template_service(template.id)
+
+        disk_attachments = self._connection.follow_link(
+            template_service.get().disk_attachments
+        )
+        disk = disk_attachments[0].disk
```
Comment (machacekondra) [inline lib/ansible/modules/cloud/ovirt/ovirt_vms.py]: Hmm, what if the template will have more disks. We have to somehow distinguish which one should go where...

Comment (tareqalayan) [reply lib/ansible/modules/cloud/ovirt/ovirt_vms.py]: I think they all should go there, no?

File: lib/ansible/modules/cloud/ovirt/ovirt_vms.py (hunk 1 above)
Comment (tareqalayan) [reply lib/ansible/modules/cloud/ovirt/ovirt_vms.py]: ok will change this

File: lib/ansible/modules/cloud/ovirt/ovirt_vms.py (hunk 2 above)
Comment (tareqalayan) [reply lib/ansible/modules/cloud/ovirt/ovirt_vms.py]: done

**Comment by tareqalayan** [comment]
@machacekondra please review 

thanks,

**Comment by machacekondra** [comment]
We need to think about the possibility of specification of specific storage of specific disk.

**Comment by machacekondra** [comment]
Maybe we can do following:

Specific disk to specific SD:
```yaml
ovirt_vms:
  name: myvm
  template: mytemplate
  cluster: mycluster
  disks:
    -  id: 123
       storage_domain: nfs_0
```

All disks to one SD:
```yaml
ovirt_vms:
  name: myvm
  template: mytemplate
  cluster: mycluster
  disks:
    - storage_domain: nfs_0
```
What do you think, any better idea?

**Comment by tareqalayan** [comment]
@machacekondra 
OK sounds good, 
I will do the change during the week or beginning of next week.

**Comment by mattclay** [comment]
Closing and re-opening to trigger CI.

**Comment by jctanner** [comment]
bot_status

**Comment by tareqalayan** [comment]
@machacekondra 

i think if there is template in the module we don't have to specify disk_id or name we just need to copy all disks attached to template to the specified storage domain? 

i am just thinking loud here ... what do you think?

**Comment by tareqalayan** [comment]
@machacekondra 
Ondra can you please review?

File: lib/ansible/modules/cloud/ovirt/ovirt_vms.py [hunk 3]
```diff
@@ -477,8 +478,55 @@ def __get_template_with_version(self):
 
         return template
 
+    def __get_storage_domain(self, disk_id):
+        storage_domain = None
+
```
Comment (myakove) [inline lib/ansible/modules/cloud/ovirt/ovirt_vms.py:483]: disks = self._module.params.get('disks', list())

File: lib/ansible/modules/cloud/ovirt/ovirt_vms.py [hunk 4]
```diff
@@ -477,8 +478,55 @@ def __get_template_with_version(self):
...
+        if self._module.params['disks']:
+            for disk in self._module.params['disks']:
```
Comment (myakove) [inline lib/ansible/modules/cloud/ovirt/ovirt_vms.py]: for disk in disks:

File: lib/ansible/modules/cloud/ovirt/ovirt_vms.py [hunk 5]
```diff
@@ -477,8 +478,55 @@ def __get_template_with_version(self):
...
+                        self._connection.system_service().storage_domains_service(),
+                        # self._module.params['disks'][0]['storage_domain']
+                        disk['storage_domain']
+                    )
+
+        return storage_domain
+
+    def __get_disk_attachments(self, template):
+
+        # if self.__get_storage_domain() is None:
```
Comment (myakove) [inline lib/ansible/modules/cloud/ovirt/ovirt_vms.py]: remove comments please

File: lib/ansible/modules/cloud/ovirt/ovirt_vms.py [hunk 6]
```diff
@@ -539,6 +587,7 @@ def build_entity(self):
                 self.param('serial_policy') is not None or
                 self.param('serial_policy_value') is not None
             ) else None,
+            disk_attachments=disks if disks else None,
```
Comment (myakove) [inline lib/ansible/modules/cloud/ovirt/ovirt_vms.py:587]: disks or None

File: lib/ansible/modules/cloud/ovirt/ovirt_vms.py (hunk 3 above)
Comment (tareqalayan) [reply lib/ansible/modules/cloud/ovirt/ovirt_vms.py:483]: done

File: lib/ansible/modules/cloud/ovirt/ovirt_vms.py (hunk 4 above)
Comment (tareqalayan) [reply lib/ansible/modules/cloud/ovirt/ovirt_vms.py]: done

File: lib/ansible/modules/cloud/ovirt/ovirt_vms.py (hunk 5 above)
Comment (tareqalayan) [reply lib/ansible/modules/cloud/ovirt/ovirt_vms.py]: done

File: lib/ansible/modules/cloud/ovirt/ovirt_vms.py (hunk 6 above)
Comment (tareqalayan) [reply lib/ansible/modules/cloud/ovirt/ovirt_vms.py:587]: I will leave this as is to be like the lines above.

File: lib/ansible/modules/cloud/ovirt/ovirt_vms.py [hunk 7]
```diff
@@ -477,8 +478,52 @@ def __get_template_with_version(self):
...
+
+        disks = list()
+
+        template_service = self._connection.system_service().templates_service()
+
+        template_obj = template_service.template_service(template.id)
+
+        disk_attachments = self._connection.follow_link(
+            template_obj.get().disk_attachments
+        )
```
Comment (machacekondra) [inline lib/ansible/modules/cloud/ovirt/ovirt_vms.py:505]: No need to re-fetch the template, you can just write:
```python
disk_attachments = self._connection.follow_link(template.disk_attachments)
```

File: lib/ansible/modules/cloud/ovirt/ovirt_vms.py [hunk 8]
```diff
@@ -477,8 +478,52 @@ def __get_template_with_version(self):
...
+
+        for disk in disk_attachments:
+            disks.append(
+                otypes.DiskAttachment(
+                    disk=otypes.Disk(
+                        id=disk.id,
+                        format=otypes.DiskFormat.COW,
```
Comment (machacekondra) [inline lib/ansible/modules/cloud/ovirt/ovirt_vms.py:512]: What about 'raw'?

File: lib/ansible/modules/cloud/ovirt/ovirt_vms.py [hunk 9]
```diff
@@ -477,8 +478,52 @@ def __get_template_with_version(self):
...
         return template
 
+    def __get_storage_domain(self, disk_id):
+        storage_domain = None
+
+        disks = self._module.params.get('disks', list())
+
+        for disk in disks:
+            if disk_id == disk.get('id'):
+                storage_domain = search_by_name(
```
Comment (machacekondra) [inline lib/ansible/modules/cloud/ovirt/ovirt_vms.py:488]: Better to use ```get_id_by_name``` it get proper error message if storage is not found

File: lib/ansible/modules/cloud/ovirt/ovirt_vms.py (hunk 6 above)
Comment (machacekondra) [reply lib/ansible/modules/cloud/ovirt/ovirt_vms.py:587]: I wonder what should happen when disks is [].

File: lib/ansible/modules/cloud/ovirt/ovirt_vms.py [hunk 10]
```diff
@@ -709,6 +755,9 @@ def wait_for_down(self, vm):
     def __attach_disks(self, entity):
         disks_service = self._connection.system_service().disks_service()
 
+        if self.param('template'):
```
Comment (machacekondra) [inline lib/ansible/modules/cloud/ovirt/ovirt_vms.py:758]: What if we want to create VM from template, and attach some other disks?

File: lib/ansible/modules/cloud/ovirt/ovirt_vms.py [hunk 11]
```diff
@@ -477,8 +478,52 @@ def __get_template_with_version(self):
 
         return template
 
+    def __get_storage_domain(self, disk_id):
+        storage_domain = None
+
+        disks = self._module.params.get('disks', list())
+
+        for disk in disks:
+            if disk_id == disk.get('id'):
```
Comment (machacekondra) [inline lib/ansible/modules/cloud/ovirt/ovirt_vms.py:487]: Can we somehow implement behavior of specify storage_domain for all disks? If I understand the code correctly, it not currently.

**Review by machacekondra — CHANGES_REQUESTED** [review]
Can you also verify the beahvior of copy/move disks?

File: lib/ansible/modules/cloud/ovirt/ovirt_vms.py (hunk 11 above)
Comment (tareqalayan) [reply lib/ansible/modules/cloud/ovirt/ovirt_vms.py:487]: The idea here is to specify which disk go where see example:
  ovirt_vms:
    name: vm_test
    cluster: clu1
    template: template01
    operating_system: other_linux
    type: server
    cpu_cores: 1
    cpu_sockets: 1
    state: stopped
    clone: True
    timeout: 600
    disks:
      - id: disk_id1
        storage_domain: my_nfs_1
     -  id: disk_id1
        storage_domains: my_nfs_3

i would prefer not to decalre any disk and only storage domain so all disks related to template will go to spesfic storage domain ... 
but you said there is a use case to split disks of templates on multiple storage domains .... 

please tell me what you suggest... and i will implement it ...

File: lib/ansible/modules/cloud/ovirt/ovirt_vms.py (hunk 9 above)
Comment (tareqalayan) [reply lib/ansible/modules/cloud/ovirt/ovirt_vms.py:488]: ack

File: lib/ansible/modules/cloud/ovirt/ovirt_vms.py (hunk 7 above)
Comment (tareqalayan) [reply lib/ansible/modules/cloud/ovirt/ovirt_vms.py:505]: ack

File: lib/ansible/modules/cloud/ovirt/ovirt_vms.py (hunk 8 above)
Comment (tareqalayan) [reply lib/ansible/modules/cloud/ovirt/ovirt_vms.py:512]: I didn't put much thinking here ... what it should it be?

File: lib/ansible/modules/cloud/ovirt/ovirt_vms.py (hunk 6 above)
Comment (tareqalayan) [reply lib/ansible/modules/cloud/ovirt/ovirt_vms.py:587]: nothing should happen it put the disk on the first storage storage domain on the list ... in my case it always choose nfs_0 ....

File: lib/ansible/modules/cloud/ovirt/ovirt_vms.py (hunk 10 above)
Comment (tareqalayan) [reply lib/ansible/modules/cloud/ovirt/ovirt_vms.py:758]: i am not familiar with this use case. 

do we want to support that ? 

should i  grab all template disks + disks stated in the module?

File: lib/ansible/modules/cloud/ovirt/ovirt_vms.py (hunk 11 above)
Comment (tareqalayan) [reply lib/ansible/modules/cloud/ovirt/ovirt_vms.py:487]: the plan now is: 
1. if there is a template specified and a storage domain: all disks related to that template will go into the storage domain
2. if there is a template specifeid and disks is specified each disk will go to the specified storage domain:
template: my_template
disks:
  - id_of_disk
  - storage domain 


@machacekondra please ack

**Pull request closed without merging**